Approved/remove-new-array.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Remove default-initialized new array expressions
+description: The `new Type[N]` array expression is deprecated and removed from Q#.
+author: Sarah Marshall
+date: October 28, 2020
+---
+
+# Proposal
+
+The `new Type[N]` expression, which creates an array of `N` values using the default value of `Type`, is deprecated and removed from Q#.
+
+# Justification
+
+The existence of `new Type[N]` has negative effects on the language.
+
+The most important negative effect is the assumption that every type has a default value.
+This is not a reasonable assumption, because it is not possible to define useful default values of types like `Qubit` and `a -> b`.
+Their current default values are invalid and will trigger an error if they are used.
+But the syntax is the same both for types with valid default values, and those without, which is misleading: while `new Int[3]` is a well-defined value whose items can be used safely, `new Qubit[3]` creates a potential trap that will crash the program if any item in the array is used before being set to a valid value first.
+
+This assumption is also invalid in the presence of uninhabited types such as `Void`.
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+Finally, the use of the `new` keyword is inconsistent with other ways to create values in Q#.
+`new` is required only when creating a default-initialized array, not when using a square bracket array literal or when creating values of other types.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be removed.
+The same functionality can be accomplished with functions in the standard library:
+
+```qsharp
+// Create an empty arra of type Int[].
+let empty = EmptyArray<Int>();
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = ConstantArray(10, false);
+```
+
+# Implementation
+
+The library functions `EmptyArray` and `ConstantArray` are already implemented.
+The `new Type[N]` syntax can be deprecated and removed with the next major version of Q#.
+
+## Timeline
+
+N/A
+
+# Further Considerations
+
+## Related Mechanisms
+
+N/A
+
+## Impact on Existing Mechanisms
+
+N/A
+
+## Anticipated Interactions with Future Modifications
+
+Improved type inference will allow the empty array to be expressed as `[]` instead of `EmptyArray<a>()`.
+
+## Alternatives
+
+N/A
```

Comment:
I think it is worthwhile to consider whether it makes sense to introduce syntax sugar for common array instantiations, such as ConstantArray. I think this might make sense here.

Reply:
Do you have a concrete example of syntax for a constant array? I can't think of anything reasonable.

Reply:
I agree we don't need to rush it. It likely makes more sense to first introduce multi-dimensional arrays. Then we can consider if and what array manipulation capabilities we would like to support in general, and what syntax sugar might make sense. I think what I was thinking was somewhat aligned with having an operator that given an array, constructs a stack or sequence of x such arrays, but I agree that doesn't necessarily seem that useful more generally. It would still be good to add a sentence here that adding dedicated syntax is something one could consider at this point, but we have decided that it makes more sense to potentially support it later only if there is a more general need for array manipulation operators.

Reply:
Link to multidimensional arrays proposal: https://github.com/microsoft/qsharp-language/pull/49